### PR TITLE
feat: allow custom shell opts for run-script

### PR DESCRIPTION
**Pull request by madarche**
Usage:
```shell
$ npm config set script-shell /bin/bash
$ npm config set script-shell-opts "-o pipefail -c"
```

This PR is related to #16687 

The aim of this PR is to be able to specify custom shell options such as "-o pipefail" for bash or any other shell, so that npm can return error on pipe-chained commands which are very common when one uses npm as a build tool with lots of npm scripts.

### Example

```json
"scripts": {
    "build:js": "browserify -g envify -e www/js/main.js | uglifyjs -c -m | tee www/bundle/bundle.js | gzip > www/bundle/bundle.js.gz",
  }
```

The above `browserify` command may fail, but the other piped commands (`uglifyjs`, `tee`, `gzip`) won't and thus the `build:js` script won't fail as a whole and a CI will consider the `npm run build:js` a success and will proceed to the next action.

If `npm config set script-shell-opts "-o pipefail -c"` was supported, the `build:js` npm script will produce an error and the CI will fail, providing a much safer situation.

### Reference

And for information here is [another reference of `pipefail` usage on stackoverflow](https://stackoverflow.com/questions/4472318/error-handling-with-chained-commands-pipes-in-a-bash-script).

Cheers

### CODE REVIEW

File: lib/utils/lifecycle.js [hunk 1]
```diff
@@ -265,9 +265,11 @@ function runCmd_ (cmd, pkg, env, wd, stage, unsafe, uid, gid, cb_) {
   var shFlag = '-c'
 
   var customShell = npm.config.get('script-shell')
+  var customShellOpts = npm.config.get('script-shell-opts')
 
   if (customShell) {
     sh = customShell
+    shFlag += ' ' + customShellOpts
```
Comment (mmkal) [inline lib/utils/lifecycle.js]: shouldn't this just be `shFlag = customShellOpts`? It's conceivable that you wouldn't want `-c` used at all.

**Comment by madarche** [comment]
@mmkal thanks for your interest and your remark. Please tell me if you agree with the current state of this PR.

**Comment by madarche** [comment]
PR rebased onto `latest`

**Comment by danielbayley** [comment]
Would it not be better to simply allow `npm config set script-shell '/bin/bash -o pipefail -c'`?

**Comment by madarche** [comment]
@danielbayley: Doing `npm config set script-shell '/bin/bash -o pipefail -c'` doesn't work (please try it yourself) because of the way the shell is spawned:
```javascript
var proc = spawn(sh, [shFlag, cmd], conf, opts.log)
```

That's why I've proposed this PR.

**Comment by zkat** [comment]
Hi! We're moving repos to https://github.com/npm/cli/pulls! See [our blog post about the migration to npm.community](https://blog.npmjs.org/post/175587538995/announcing-npmcommunity) for details. As such, we're closing all active PRs on this repo.

Could you please re-open this PR against `npm/cli` instead? We're still interested in this patch so we hope you will! We'll continue discussion over there once it's moved. 💚

**Pull request closed without merging**